runtime-manager/src/runtime_manager_icecap.rs

```diff
@@ -0,0 +1,203 @@
+//! IceCap-specific material for the Runtime Manager enclave
+//!
+//! ## Authors
+//!
+//! The Veracruz Development Team.
+//!
+//! ## Licensing and copyright notice
+//!
+//! See the `LICENSE.markdown` file in the Veracruz root directory for
+//! information on licensing and copyright.
+
+extern crate alloc;
+
+use std::string::ToString;
+use serde::{Serialize, Deserialize};
+use bincode::{serialize, deserialize};
+
+use icecap_core::prelude::*;
+use icecap_core::config::RingBufferConfig;
+use icecap_core::logger::{Logger, Level, DisplayMode};
+use icecap_start_generic::declare_generic_main;
+
+use veracruz_utils::platform::icecap::message::{Request, Response, Error};
+
+use crate::managers::session_manager;
+
+declare_generic_main!(main);
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct Config {
+    host_ring_buffer: RingBufferConfig,
+}
+
+fn main(config: Config) -> Fallible<()> {
+    icecap_runtime_init();
+    let host_ring_buffer = RingBuffer::realize_resume(&config.host_ring_buffer);
+    let host_ring_buffer_notification = config.host_ring_buffer.wait;
+    RuntimeManager::new(host_ring_buffer, host_ring_buffer_notification).run()
+}
+
+struct RuntimeManager {
+    host_ring_buffer: PacketRingBuffer,
+    host_ring_buffer_notification: Notification,
+    active: bool,
+}
+
+impl RuntimeManager {
+
+    fn new(host_ring_buffer: RingBuffer, host_ring_buffer_notification: Notification) -> Self {
+        host_ring_buffer.enable_notify_read();
+        host_ring_buffer.enable_notify_write();
+        Self {
+            host_ring_buffer: PacketRingBuffer::new(host_ring_buffer),
+            host_ring_buffer_notification,
+            active: true,
+        }
+    }
+
+    fn run(&mut self) -> Fallible<()> {
+        loop {
+            let req = self.recv()?;
+            let resp = self.handle(&req)?;
+            self.send(&resp)?;
+            if !self.active {
+                std::icecap_impl::external::runtime::exit();
+            }
+        }
+    }
+
+    fn handle(&mut self, req: &Request) -> Fallible<Response> {
+        Ok(match req {
+            Request::New { policy_json } => {
+                session_manager::init_session_manager(&policy_json).unwrap();
+                Response::New
+            }
+            Request::GetEnclaveCert => {
+                match session_manager::get_enclave_cert_pem() {
+                    Err(s) => {
+                        log::debug!("{}", s);
```

Comment:
I find it helpful to leave some logging statements in, so that, when debugging, one can go into this source file and set the log level to include them. By default, the log level within the enclave is low enough that none are actually printed.